File: apps/web/src/stores/bookmarks.ts
import { createWithSignal } from "solid-zustand";

import { Invoke } from "../invokes";
import type { Bookmark, FolderData } from "../types";

type BookmarkState = {
  folders: FolderData[];
  bookmarks: Bookmark;
  getFolders: () => void;
  getBookmarks: (index: number) => void;
  addBookmark: (url: string, title: string) => void;
  removeBookmark: (index: number) => void;
};

export const useBookmarkState = createWithSignal<BookmarkState>((set) => ({
  folders: [],
  bookmarks: {
    index: 1,
    title: "All Bookmarks",
    url: null,
    host: null,
    node_type: "Root",
    date_added: 0,
    children: [],
  },
  getFolders: async () => {
    const folders = await Invoke.GetRootChildrenFolder();
    set(() => ({ folders }));
  },
  getBookmarks: async (index) => {
    // can't accept index 0 because indextree starts from 1
    if (index >= 1) {
      const data = await Invoke.GetNestedJson(index);
      const tree = JSON.parse(data) as Bookmark;
      set(() => ({ bookmarks: tree }));
    }
  },
  addBookmark: async (url, title) => {
    // get current top of bookmark index that shown in the UI as a starting point
    const current = useBookmarkState((state) => state.bookmarks);
    const startingIndex = current().index;
    const data = await Invoke.AddBookmark(url, title, startingIndex);
    const tree = JSON.parse(data) as Bookmark;
    set(() => ({ bookmarks: tree }));
    // update the folders list
    useBookmarkState.getState().getFolders();
  },
  removeBookmark: async (index) => {
    // get current top of bookmark index that shown in the UI as a starting point
    const current = useBookmarkState((state) => state.bookmarks);
    const startingIndex = current().index;
    const data = await Invoke.RemoveBookmark(index, startingIndex);
    const tree = JSON.parse(data) as Bookmark;
    set(() => ({ bookmarks: tree }));
    // update the folders list
    useBookmarkState.getState().getFolders();
  },
}));

File: src-tauri/src/commands/bookmarks.rs
use std::sync::Mutex;

use mejiro_core::{bookmarks::Bookmarks, data::FolderData};

use crate::error::AppError;

#[tauri::command]
pub async fn get_nested_json(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
) -> Result<String, AppError> {
    if index == 0 {
        return Err(AppError::Other("index should not be 0".to_string()));
    }
    let bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    Ok(bookmarks.to_nested_json(index)?)
}

#[tauri::command]
pub fn get_root_children_folder(
    state: tauri::State<'_, Mutex<Bookmarks>>,
) -> Result<Vec<FolderData>, AppError> {
    let bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    let children = bookmarks.get_root_children_folder()?;
    Ok(children)
}

#[tauri::command]
pub async fn add_bookmark(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    url: String,
    title: Option<String>,
    starting_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.add_bookmark(url, title, starting_index)?;

    Ok(bookmarks.to_nested_json(starting_index)?)
}

#[tauri::command]
pub fn remove_bookmark(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
    starting_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.remove_subtree(index)?;

    Ok(bookmarks.to_nested_json(starting_index)?)
}
